src/main/java/io/fusionauth/http/security/SecurityTools.java

```diff
@@ -75,25 +80,93 @@ public static SSLContext clientContext(Certificate certificate) throws GeneralSe
+    var is = new ByteArrayInputStream(certificateString.getBytes());
+    var certs = (Collection<X509Certificate>) factory.generateCertificates(is);
+    return reorderCertificates(certs);
   }
 
-  public static byte[] parseDERFromPEM(String pem, String beginDelimiter, String endDelimiter) {
-    int startIndex = pem.indexOf(beginDelimiter);
-    if (startIndex < 0) {
-      throw new IllegalArgumentException("Invalid PEM format");
+  /**
+   * Returns an array of Certificates ordered starting from the end-entity through each supplied issuer in the List.
+   */
+  private static Certificate[] reorderCertificates(Collection<X509Certificate> certs) {
+    if (certs.isEmpty()) {
+      throw new IllegalArgumentException("Empty certificate list");
     }
 
-    int endIndex = pem.indexOf(endDelimiter);
-    if (endIndex < 0) {
-      throw new IllegalArgumentException("Invalid PEM format");
+    var orderedCerts = new X509Certificate[certs.size()];
+
+    // Short circuit if only one cert is in collection
+    if (certs.size() == 1) {
+      orderedCerts[0] = certs.stream().findFirst().get();
+      return orderedCerts;
+    }
+
+    // Index certificates by Issuer and Subject
+    var certsByIssuer = new HashMap<X500Principal, X509Certificate>(certs.size());
+    var certsBySubject = new HashMap<X500Principal, X509Certificate>(certs.size());
+
+    // Load all certificates into maps
+    for (X509Certificate cert : certs) {
+      certsByIssuer.put(cert.getIssuerX500Principal(), cert);
+      certsBySubject.put(cert.getSubjectX500Principal(), cert);
+    }
+
+    // Find the server certificate. It's the one that no other certificate refers to as its issuer. Store it in the first array element.
+    for (var cert : certs) {
+      if (!certsByIssuer.containsKey(cert.getSubjectX500Principal())) {
+        orderedCerts[0] = cert;
+        break;
+      }
+    }
+
+    // Start at the server cert and add each issuer certificate in order.
+    for (int i = 0; i < orderedCerts.length - 1; i++) {
+      var issuer = certsBySubject.get(orderedCerts[i].getIssuerX500Principal());
+      if (issuer == null) {
+        throw new IllegalArgumentException("Missing issuer cert for " + orderedCerts[i].getIssuerX500Principal());
+      }
+      orderedCerts[i + 1] = issuer;
     }
 
-    // Strip all the whitespace since the PEM and DER allow them but they aren't valid in Base 64 encoding
-    String base64 = pem.substring(startIndex + beginDelimiter.length(), endIndex).replaceAll("\\s", "");
-    return Base64.getDecoder().decode(base64);
+    return orderedCerts;
+  }
+
+
+  /**
+   * Parses all objects in a PEM-formatted string into a byte[].
+   */
+  public static byte[] parseDERFromPEM(String pem, String beginDelimiter, String endDelimiter) {
+    // Allocate an initial buffer to hold 2 certificates.
+    ByteArrayOutputStream outBytes = new ByteArrayOutputStream(4000);
+
+    try {
+      while (!pem.trim().isEmpty()) {
```

Comment:
The changes here aren't strictly necessary anymore, since CertificateFactory handles loading multiple PEM entries from a single stream now. Remove?

Reply:
not sure if I follow all of the logic here, but if we don't need it, I am always in favor of deleting code. :-)

Reply:
Reverted to previous code.